<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 4: West Nile Virus Prediction

# Part 1 of 4

# Executive Summary 
Insert text

## Problem Statement
Insert text

## Contents: 

Part 1: Data Cleaning

-[Example](#Example)

Part 2: EDA 

Part 3: Modelling and Model Evaluation

Part 4: Recommendation and Conclusion

## Data Sets

* [`spray.csv`](./data/spray.csv): Description
* [`test.csv`](./data/test.csv): 
* [`train.csv`](./data/train.csv): 
* [`weather.csv`](./data/weather.csv): 

# Data Cleaning

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#data importing
train = pd.read_csv('../datasets/train.csv')
test = pd.read_csv('../datasets/test.csv')
spray = pd.read_csv('../datasets/spray.csv')
weather = pd.read_csv('../datasets/weather.csv')

In [3]:
#configure jupyter lab to prevent truncation for easier reference for columns
pd.set_option('display.max_columns', None)

In [4]:
train.head()

Date                                            Address  \
0  2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1  2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2  2007-05-29  6200 North Mandell Avenue, Chicago, IL 60646, USA   
3  2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   
4  2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2          CULEX RESTUANS     62    N MANDELL AVE  T007   
3  CULEX PIPIENS/RESTUANS     79     W FOSTER AVE  T015   
4          CULEX RESTUANS     79     W FOSTER AVE  T015   

              AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
1  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
2   6200  N MANDELL AVE, Chicago, IL  41.994991 -87.769279                9   
3    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   
4    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   

   NumMosquitos  WnvPresent  
0             1           0  
1             1           0  
2             1           0  
3             1           0  
4             4           0

In [5]:
test.head()

Id        Date                                            Address  \
0   1  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1   2  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2   3  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
3   4  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
4   5  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2           CULEX PIPIENS     41   N OAK PARK AVE  T002   
3        CULEX SALINARIUS     41   N OAK PARK AVE  T002   
4         CULEX TERRITANS     41   N OAK PARK AVE  T002   

              AddressNumberAndStreet  Latitude  Longitude  AddressAccuracy  
0  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
1  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
2  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
3  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
4  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9

In [6]:
spray.head()

Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [7]:
weather.head()

Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

  Sunrise Sunset CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
0    0448   1849             0      M      0.0        0.00       29.10   
1       -      -             M      M        M        0.00       29.18   
2    0447   1850      BR     0      M      0.0        0.00       29.38   
3       -      -   BR HZ     M      M        M        0.00       29.44   
4    0446   1851             0      M      0.0        0.00       29.39   

  SeaLevel  ResultSpeed  ResultDir AvgSpeed  
0    29.82          1.7         27      9.2  
1    29.82          2.7         25      9.6  
2    30.09         13.0          4     13.4  
3    30.08         13.3          2     13.4  
4    30.12         11.7          7     11.9

### Data Cleaning

#### Spray Dataset

In [9]:
train = train.groupby(by=['Date', 'Address', 'Species', 'Block', 'Street', 'Trap', 'AddressNumberAndStreet', 'Latitude', 'Longitude', 'AddressAccuracy']).agg('sum')

In [10]:
train = pd.DataFrame(train.to_records())

In [11]:
#checking
print(spray.shape)
spray.head()

(14835, 4)


Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [12]:
#convert column names to lowercase for easier referencing
spray.columns = [col.lower() for col in spray.columns]
spray.head()

date        time   latitude  longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [13]:
#check for duplicates
spray.duplicated().sum()

541

In [14]:
#drop duplicates
spray.drop_duplicates(inplace=True)

In [15]:
#check dtypes
spray.dtypes

date          object
time          object
latitude     float64
longitude    float64
dtype: object

In [16]:
#convert date to datetime object
spray['date'] = pd.to_datetime(spray['date'], format = "%Y-%m-%d")

In [17]:
#check for null values
spray.isnull().sum()

date           0
time         584
latitude       0
longitude      0
dtype: int64

In [18]:
#drop 'Time' column
spray.drop(columns='time', inplace=True)

After checking for duplicates and dropping these rows, we checked for missing values and 584 were found in the `'time'` column, which indicates the exact time the spraying was conducted. However, in our main train & test datasets, observations are only by date. Thus, we will be dropping this column. Conveniently, this removes all the null values in this dataset as well.

Then, we checked the datatypes of the dataset to see if it corresponds to their respective data. We found that `'date'` column has object datatype. Hence, we convert it to a default format of datetime.

### Weather Dataset

In [19]:
#checking
print(weather.shape)
weather.head()

(2944, 22)


Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

  Sunrise Sunset CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
0    0448   1849             0      M      0.0        0.00       29.10   
1       -      -             M      M        M        0.00       29.18   
2    0447   1850      BR     0      M      0.0        0.00       29.38   
3       -      -   BR HZ     M      M        M        0.00       29.44   
4    0446   1851             0      M      0.0        0.00       29.39   

  SeaLevel  ResultSpeed  ResultDir AvgSpeed  
0    29.82          1.7         27      9.2  
1    29.82          2.7         25      9.6  
2    30.09         13.0          4     13.4  
3    30.08         13.3          2     13.4  
4    30.12         11.7          7     11.9

In [20]:
#convert column names to lowercase for easier referencing
weather.columns = [col.lower() for col in weather.columns]
weather.head()

station        date  tmax  tmin tavg depart  dewpoint wetbulb heat cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

  sunrise sunset codesum depth water1 snowfall preciptotal stnpressure  \
0    0448   1849             0      M      0.0        0.00       29.10   
1       -      -             M      M        M        0.00       29.18   
2    0447   1850      BR     0      M      0.0        0.00       29.38   
3       -      -   BR HZ     M      M        M        0.00       29.44   
4    0446   1851             0      M      0.0        0.00       29.39   

  sealevel  resultspeed  resultdir avgspeed  
0    29.82          1.7         27      9.2  
1    29.82          2.7         25      9.6  
2    30.09         13.0          4     13.4  
3    30.08         13.3          2     13.4  
4    30.12         11.7          7     11.9

In [21]:
#check for duplicates
weather.duplicated().sum()

0

In [22]:
#check dtypes
weather.dtypes

station          int64
date            object
tmax             int64
tmin             int64
tavg            object
depart          object
dewpoint         int64
wetbulb         object
heat            object
cool            object
sunrise         object
sunset          object
codesum         object
depth           object
water1          object
snowfall        object
preciptotal     object
stnpressure     object
sealevel        object
resultspeed    float64
resultdir        int64
avgspeed        object
dtype: object

In [23]:
#convert date to datetime object
weather['date'] = pd.to_datetime(weather['date'], format = "%Y-%m-%d")

In [24]:
#check for null values
weather.isnull().sum()

station        0
date           0
tmax           0
tmin           0
tavg           0
depart         0
dewpoint       0
wetbulb        0
heat           0
cool           0
sunrise        0
sunset         0
codesum        0
depth          0
water1         0
snowfall       0
preciptotal    0
stnpressure    0
sealevel       0
resultspeed    0
resultdir      0
avgspeed       0
dtype: int64

After checking for duplicates (none was found) and datatypes, it seems like the date is classified as 'object', similar to the spray dataset. It was then converted to default format of datetime. As for missing values, it seems that there is none. However for missing values, we see from the dataset, there are non-imputed rows in some columns, for example `'codesum'`. Even though a null check was done and showed no missing values, it only checks for null values. Furthermore, upon checking the data dictionary for this dataset, missing data is also showed as:
* 'M' = Missing Data, or data that are not normally reported for the station.
* '-' = Values not available for sunrise/sunset. 
* 'T' = Trace amount of data (unable to quantify).

Thus, we must check for these imputed values and also non-imputed values (represented by `''` or `' '` in code).

In [25]:
#checking for missing data and non-imputed values
weather.isin(['M', '-', 'T', '', ' ']).sum().sort_values(ascending=False)

water1         2944
codesum        1609
sunset         1472
depth          1472
depart         1472
snowfall       1472
sunrise        1472
tavg             11
heat             11
cool             11
sealevel          9
stnpressure       4
wetbulb           4
avgspeed          3
preciptotal       2
date              0
dewpoint          0
tmin              0
tmax              0
resultspeed       0
resultdir         0
station           0
dtype: int64

In [26]:
#checking on how many stations
weather[['station']].value_counts()

station
1          1472
2          1472
dtype: int64

In [27]:
#check on water1
weather.groupby('station')['water1'].value_counts()

station  water1
1        M         1472
2        M         1472
Name: water1, dtype: int64

In [28]:
#check on codesum
weather.groupby('station')['codesum'].value_counts()

station  codesum           
1                              805
         RA                    161
         RA BR                 128
         BR                     66
         TSRA RA BR             48
                              ... 
2        TSRA RA BR HZ VCTS      1
         TSRA RA FG+ BR          1
         TSRA RA FG+ FG BR       1
         TSRA RA HZ FU           1
         TSRA RA VCTS            1
Name: codesum, Length: 140, dtype: int64

In [29]:
#check on sunset
weather.groupby('station')['sunset'].value_counts()

station  sunset
1        1931        96
         1930        56
         1929        48
         1923        32
         1925        32
                   ... 
         1900         8
         1901         8
         1905         8
         1912         8
2        -         1472
Name: sunset, Length: 119, dtype: int64

In [30]:
#check on sunrise
weather.groupby('station')['sunrise'].value_counts()

station  sunrise
1        0416        104
         0417         64
         0419         40
         0418         32
         0420         32
                    ... 
         0619          8
         0620          8
         0622          8
         0623          8
2        -          1472
Name: sunrise, Length: 122, dtype: int64

In [31]:
#check on depth
weather.groupby('station')['depth'].value_counts()

station  depth
1        0        1472
2        M        1472
Name: depth, dtype: int64

In [32]:
#check on depart
weather.groupby('station')['depart'].value_counts()

station  depart
1         2          93
         -1          84
         -2          80
          5          77
          1          76
          7          76
          3          75
          0          74
         -3          72
          4          71
          6          67
          8          59
         -5          57
         -4          56
         -6          50
          9          47
         10          46
         -8          43
         -7          30
         11          28
         12          28
         -9          25
         13          23
         -10         22
         14          22
         15          15
         16          12
         -11         10
         -12          8
         17           7
         -14          6
         18           6
         -13          5
         19           4
         20           4
         -15          3
         -16          3
         22           3
         -17          2
         21           2
         23           1


From the data dictionary, the first tell-tale sign for the reason for such a high rate of missing values/data is the difference in data extraction from different stations. AWOS stations will often have less data (i.e., fewer weather elements) than ASOS station. Also, it is also stated that "not all stations report snow totals" in the documentation. Since this was the case, we groupby the columns with the high missing values to the `'station'` column. From there, we can see that station 2 has most of the missing values. These are the next steps in handling the columns with the large missing values:

1) Dropping these features:
* `'water1'`, `'depth'` and `'snowfall'` as the they either consist of all missing data or are imputed as 0 value which does not provide any meaningful information.
* `'sunset'`, `'sunrise'` and `'codesum'` as weather and time will not be a focus in our project.
* `'depart'` as departure of temperature from historical norm is unlikely to affect mosquitos population.
* `'heat'` and `'cool'` as these variables are more relevant for calculating energy needs for electrical, natural gas, power, and heating, and air conditioning industries, i.e. not as relevant to risk of mosquito breeding.
* `'wetbulb'` as `'dewpoint'` will be the effective proxy of the humidity conditions.

2) Impute `'avgspeed'` with observations from other stations and impute mean of `'tmax'` & `'tmin'` for `'tavg'` feature.

3) Impute `'stnpressure'` and `'sealevel'` from previous and next day values as not much will change.

4) Dropping rows of `'preciptotal'` since data accounts for less than 1% of dataset.

In [33]:
# 1: dropping water1, depth and snowfall
weather.drop(columns=['water1', 'depth', 'snowfall','sunrise', 'sunset', 'codesum', 'depart', 'heat','cool','wetbulb'], inplace=True)

In [34]:
#impute tavg and avgspeed
def impute_tavg_avgspeed(row):
    #impute tavg with mean of tmax & tmin
    if row['tavg'] == 'M':
        row['tavg'] = round((row['tmax'] + row['tmin']) / 2)
    
    #impute avgspeed with observations from other 
    if row['avgspeed'] == 'M':
        row['avgspeed'] = weather.loc[
            (weather['date'] == row['date']) & 
            (weather['station'] == 1), 
            'avgspeed'].values[0]
    return row

In [35]:
weather = weather.apply(impute_tavg_avgspeed, axis=1)

In [36]:
#impute stnpressure and sealevel 
def impute_values(df):
    #impute stnpressure by interpolating from previous & next day values
    for index, row in df.loc[df['stnpressure'].isin(['M'])].iterrows():
        inter = (float(df.iloc[(index - 2)]['stnpressure']) + \
                 float(df.iloc[(index + 2)]['stnpressure'])) / 2
        df.at[index, 'stnpressure'] = round(inter, 2)

    #impute sealevel by interpolating from previous & next day values
    for index, row in df.loc[df['sealevel'].isin(['M'])].iterrows():
        inter = (float(df.iloc[(index - 2)]['sealevel']) + \
                 float(df.iloc[(index + 2)]['sealevel'])) / 2
        df.at[index, 'sealevel'] = round(inter, 2)    

In [37]:
impute_values(weather)

In [38]:
#dropping rows with missing values in preciptotal
weather = weather.drop(weather[(weather.preciptotal == 'M') | (weather.preciptotal == 'T')].index)

#checking
print(weather.shape)
weather

(2942, 12)


station       date  tmax  tmin tavg  dewpoint preciptotal stnpressure  \
0           1 2007-05-01    83    50   67        51        0.00       29.10   
1           2 2007-05-01    84    52   68        51        0.00       29.18   
2           1 2007-05-02    59    42   51        42        0.00       29.38   
3           2 2007-05-02    60    43   52        42        0.00       29.44   
4           1 2007-05-03    66    46   56        40        0.00       29.39   
...       ...        ...   ...   ...  ...       ...         ...         ...   
2939        2 2014-10-29    49    40   45        34        0.00       29.42   
2940        1 2014-10-30    51    32   42        34        0.00       29.34   
2941        2 2014-10-30    53    37   45        35           T       29.41   
2942        1 2014-10-31    47    33   40        25        0.03       29.49   
2943        2 2014-10-31    49    34   42        29        0.04       29.54   

     sealevel  resultspeed  resultdir avgspeed  
0       29.82          1.7         27      9.2  
1       29.82          2.7         25      9.6  
2       30.09         13.0          4     13.4  
3       30.08         13.3          2     13.4  
4       30.12         11.7          7     11.9  
...       ...          ...        ...      ...  
2939    30.07          8.5         29      9.0  
2940    30.09          5.1         24      5.5  
2941    30.10          5.9         23      6.5  
2942    30.20         22.6         34     22.9  
2943    30.20         21.7         34     22.6  

[2942 rows x 12 columns]

In [39]:
#only 2 rows dropped but T values still present
#checking for other possible T combination
weather.isin(['  T']).sum().sort_values(ascending=False)

preciptotal    318
station          0
date             0
tmax             0
tmin             0
tavg             0
dewpoint         0
stnpressure      0
sealevel         0
resultspeed      0
resultdir        0
avgspeed         0
dtype: int64

In [40]:
# Change trace values for PrecipTotal to 0.01
weather['preciptotal'] = weather['preciptotal'].map(lambda x: 0.01 if x == '  T' else x)

In [41]:
#checking
weather.isin(['M', '-', 'T', '', ' ', '  T']).sum().sort_values(ascending=False)

station        0
date           0
tmax           0
tmin           0
tavg           0
dewpoint       0
preciptotal    0
stnpressure    0
sealevel       0
resultspeed    0
resultdir      0
avgspeed       0
dtype: int64

Upon dropping the stated features and imputing the values accordingly, we ran into a puzzling situation when attempting to drop rows containing the missing values in `'preciptotal'` as even after dropping them, T values are still present. After checking for possible combination of strings with T, it was found that the values imputed were actually `'  T'` and it accounted for 318 of the data. Since precipitation is a valuable feature to hold, we decided to impute the trace amount with 0.01 as it would signify a quantity that is very low and close to zero but still shows some amount of data collected that may be meaningful in our project.

# EDA

# Modelling and Model Evaluation

# Recommendation and Conclusion